x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java

```diff
@@ -67,9 +79,48 @@
         Setting.Property.NodeScope
     );
 
-    public static final Setting<ByteSizeValue> SNAPSHOT_CACHE_SIZE_SETTING = Setting.byteSizeSetting(
+    public static final Setting<ByteSizeValue> SNAPSHOT_CACHE_SIZE_SETTING = new Setting<>(
         SHARED_CACHE_SETTINGS_PREFIX + "size",
-        ByteSizeValue.ZERO,
+        ByteSizeValue.ZERO.getStringRep(),
+        s -> ByteSizeValue.parseBytesSizeValue(s, SHARED_CACHE_SETTINGS_PREFIX + "size"),
+        new Setting.Validator<ByteSizeValue>() {
+
+            @Override
+            public void validate(final ByteSizeValue value) {
+
+            }
+
+            @Override
+            public void validate(final ByteSizeValue value, final Map<Setting<?>, Object> settings) {
+                if (value.getBytes() > 0) {
+                    @SuppressWarnings("unchecked") final List<DiscoveryNodeRole> roles =
+                        (List<DiscoveryNodeRole>) settings.get(NodeRoleSettings.NODE_ROLES_SETTING);
+                    final DiscoveryNode fake = new DiscoveryNode(
+                        "fake",
+                        new TransportAddress(TransportAddress.META_ADDRESS, 0),
+                        Map.of(),
+                        Set.of(roles.toArray(DiscoveryNodeRole[]::new)), Version.CURRENT
+                    );
+                    if (DataTier.isFrozenNode(fake) == false) {
+                        deprecationLogger.deprecate(
+                            DeprecationCategory.SETTINGS,
+                            "shared_cache",
+                            "setting [{}] to be positive [{}] on node without the data_frozen role is deprecated, roles are [{}]",
+                            SHARED_CACHE_SETTINGS_PREFIX + "size",
+                            value.getStringRep(),
+                            roles.stream().map(DiscoveryNodeRole::roleName).collect(Collectors.toUnmodifiableList())
```

Comment:
Perhaps
```suggestion
                            roles.stream().map(DiscoveryNodeRole::roleName).collect(Collectors.joining(","))
```
to avoid the double `[[]]` wrapping